python/ql/src/semmle/python/frameworks/Stdlib.qll

```diff
@@ -864,6 +864,212 @@ private module Stdlib {
+          "lchmod", "lstat", "mkdir", "open", "owner", "read_bytes", "read_text", "readlink",
+          "rename", "replace", "resolve", "rglob", "rmdir", "samefile", "symlink_to", "touch",
+          "unlink", "link_to", "write_bytes", "write_text"
+        ] and
+      fileAccess.getObject().getALocalSource() = pathlibPath() and
+      this.getFunction() = fileAccess
+    }
+
+    override DataFlow::Node getAPathArgument() { result = fileAccess.getObject() }
+  }
+
+  /** An additional taint steps for objects of type `pathlib.Path` */
+  private class PathlibPathTaintStep extends TaintTracking::AdditionalTaintStep {
+    override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+      // Type construction
+      nodeTo = pathlib().getMember(pathlibPathConstructor()).getACall() and
+      nodeFrom = nodeTo.(DataFlow::CallCfgNode).getArg(_)
+      or
+      // Type preservation
+      nodeFrom.getALocalSource() = pathlibPath() and
+      nodeTo.getALocalSource() = pathlibPath() and
+      (
+        // Type-preserving call
+        typePreservingCall(nodeFrom, nodeTo)
+        or
+        // Type-preserving attribute access
+        typePreservingAttribute(nodeFrom, nodeTo)
+      )
+      or
+      // Data injection
+      nodeTo.getALocalSource() = pathlibPath() and
+      (
+        // Special handling of the `/` operator
+        exists(BinaryExprNode slash, DataFlow::Node pathOperand |
+          slash.getOp() instanceof Div and
+          pathOperand.asCfgNode() = slash.getAnOperand() and
+          pathOperand.getALocalSource() = pathlibPath()
+        |
+          nodeTo.asCfgNode() = slash and
+          // Taint can flow either from the left or the right operand as long as one of them is a path.
+          nodeFrom.asCfgNode() = slash.getAnOperand()
+        )
+        or
+        // standard case
+        exists(DataFlow::AttrRead augmentsPath |
+          augmentsPath.getAttributeName() = pathlibPathInjection()
+        |
+          nodeTo.(DataFlow::CallCfgNode).getFunction() = augmentsPath and
+          (
+            // type-preserving call
+            nodeFrom = augmentsPath.getObject()
+            or
+            // data injection
+            nodeFrom = nodeTo.(DataFlow::CallCfgNode).getArg(_)
+          )
+        )
+      )
+      or
+      // Export data from type
+      nodeFrom.getALocalSource() = pathlibPath() and
+      (
+        // exporting attribute
+        exists(DataFlow::AttrRead export |
+          export.getAttributeName() = pathlibPathAttributeExport()
+        |
+          nodeTo = export and
+          nodeFrom = export.getObject()
+        )
+        or
+        // exporting call
+        exists(DataFlow::AttrRead export | export.getAttributeName() = pathlibPathMethodExport() |
+          nodeTo.(DataFlow::CallCfgNode).getFunction() = export and
+          nodeFrom = export.getObject()
+        )
```

Comment:
Hmm... This sort of construction won't handle cases where `foo.bar()` is split up as `x = foo.bar; x()`. I think we should consider adding some local flow internally to handle this. I guess really `AttrRead` should be a `LocalSourceNode`, and then you could use `flowsTo` to link it up.

Reply:
So you would add an exception for `AttrRead`s similar to the one for `any(ModuleVariableNode mvn).getARead()`?

Reply:
(in the charpred for `LocalSourceNode`)

Reply:
Nah, what I meant was that it ought to be a `LocalSourceNode` already. We just don't witness this in the QL types.

Reply:
So you would not expect to have local flow into attributes?

Reply:
That really depends on what you mean by "into attributes". For `foo.bar = baz`, sure there's flow into `foo.bar` (but this is an `AttrWrite`). For `x = foo.bar + baz`, say, the `foo.bar` component is only a _source_ of flow (though in this case it only flows to itself). Of course there's flow to the `foo` part of both of these.

Reply:
So in the program
```python
def the_local_scope(baz):
  foo.bar = baz
  x = foo.bar + baz
  return x
```
would you think that the `AttrRead` should be a `LocalSourceNode`? I guess in this case, where `foo` is apparently global, any read should probably be considered a new value. But in general, I do not see why reads should be local sources?

Reply:
Let's ignore the "apparently global" issue for now by making `foo` an argument as well:
```python
def the_local_scope(foo, baz):
    foo.bar = baz
    x = foo.bar + baz
    return x
```
What are the local sources above? Let me count them:
1. `foo` as it appears in the list of parameters to this function. It flows to all other occurrences of `foo`.
2. `baz` likewise.
3. `foo.bar` as it appears in the penultimate line. It flows to only itself.
4. `foo.bar + baz`, which flows to all occurrences of `x`.

I think the confusion here is that there's some imprecision about whether we're talking about `foo.bar` in its entirety, or just the `foo` part thereof when we're talking about `AttrRead`s. To be clear: `foo.bar` is the `AttrRead`. There will (in general) be flow into the _object_ part (i.e. `foo`) of this `AttrRead`, but the entire read is a different object.

We _could_ of course add the local flow you're talking about, but for type tracking, it would be superfluous. We already have flow from `baz` to `foo.bar` (because of `foo.bar = baz`), so we track `foo has type of baz in attribute bar` from `foo` to all of its uses. This includes `foo.bar + baz`, at which point we unpack and conclude that `foo.bar` has the same type as `baz` (whatever that may be).

I'm also not sure it would be sound to add this back in as local flow. Inside a local scope, it will _mostly_ be true that `foo.bar` refers to the same object in both places (and so `x` is really `baz + baz`), but in principle `bar` could be defined using an attribute getter/setter that returns a _different_ object. So with a strict "`LocalSourceNode`s are representatives of pointer equivalence classes" view, it is incorrect to add flow here. For type trackers, on the other hand, we make the assumption that at least the _type_ (whatever its semantics) is the same, which I think is reasonable.

Finally, perhaps I should just have replaced the big wall of text above with this: https://lgtm.com/query/325540041078017890/ :smile:

Reply:
Aha, so when you say "the entire read is a different object" that is to cater for the case of a getter? And the same reasoning yields that any attribute read is a local source because it is (possibly) creating a new value? In that case we could indeed have `AttrRead` extend `LocalSourceNode` (and we could close the predicates under `flowsTo`).

For now, would you prefer that I cast to `LocalSource` and use `flowsTo`?

Reply:
I think using the cast is sufficient for now. We can always fix this up later. :+1:

Reply:
Done.